### PR TITLE
Requirements for Inference Service v0.1 RSNA 2021

### PR DESCRIPTION
**Pull request by pritishnahar95**
This PR provides a draft of the requirements for the MONAI Inference Service.

### CODE REVIEW

File: components/inference-service/requirements.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+## Overview
+As data scientists & applications developers build AI models they need a way to deploy these applications in production. MONAI Inference Service will be used to deploy a MONAI application. This proposal documents the requirements for the MONAI Inference Service (MIS).
+
+## Goal
+The goal for this proposal is to enlist, prioritize and provide clarity on the requirements for MIS. Developers working on different software modules in MIS SHALL use this specification as a guideline when designing and implementing software for the Service.
+
+## Success Criteria
+MIS SHALL enable clients to communicate with it via a REST based RPC API.
```
Comment (whoisj) [inline components/inference-service/requirements.md:8]: maybe `MIS SHALL provide a REST based RCP API for client to communicate with` ?

wording things correctly is hard 😔

File: components/inference-service/requirements.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+Clients SHALL be able to register the configuration of a [MONAI Application Packages(MAP)](https://github.com/Project-MONAI/monai-deploy/blob/main/guidelines/monai-application-package.md#manifests) with MIS.
```
Comment (whoisj) [inline components/inference-service/requirements.md:10]: I do think any of the requirements for MIS should start with `Clients SHALL`. These are requirements for MIS, not for clients.

maybe `MIS SHALL support configuration of the MONAI Application Package (MAP) used to service inference requests.`?

File: components/inference-service/requirements.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+## Goal
+The goal for this proposal is to enlist, prioritize and provide clarity on the requirements for MIS. Developers working on different software modules in MIS SHALL use this specification as a guideline when designing and implementing software for the Service.
+
+## Success Criteria
+MIS SHALL enable clients to communicate with it via a REST based RPC API.
+
+Clients SHALL be able to register the configuration of a [MONAI Application Packages(MAP)](https://github.com/Project-MONAI/monai-deploy/blob/main/guidelines/monai-application-package.md#manifests) with MIS. 
+
+Clients SHALL be able to submit a inference request to the server for running a job corresponding to a MAP along with an input dataset.
```
Comment (whoisj) [inline components/inference-service/requirements.md:12]: again, let's not `Clients SHALL` anything - we don't control them.

File: components/inference-service/requirements.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+The goal for this proposal is to enlist, prioritize and provide clarity on the requirements for MIS. Developers working on different software modules in MIS SHALL use this specification as a guideline when designing and implementing software for the Service.
+
+## Success Criteria
+MIS SHALL enable clients to communicate with it via a REST based RPC API.
+
+Clients SHALL be able to register the configuration of a [MONAI Application Packages(MAP)](https://github.com/Project-MONAI/monai-deploy/blob/main/guidelines/monai-application-package.md#manifests) with MIS. 
+
+Clients SHALL be able to submit a inference request to the server for running a job corresponding to a MAP along with an input dataset. 
+
+MIS SHALL service inference requests one at a time.
```
Comment (whoisj) [inline components/inference-service/requirements.md:14]: coming back to this, I feel like this is more of a limitation and it is not a requirement.

perhaps we should append a `## Limitations` section to this document?

File: components/inference-service/requirements.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+MIS SHALL return inference results as part of the response to the originating inference request.
+
+## Requirements
+
+### Support for Specific MONAI Workloads
+MIS SHALL support [MONAI workloads](https://github.com/Project-MONAI/monai-deploy/blob/main/guidelines/monai-workloads.md#clinical-use-cases) which can be completed within the timeframe of a single HTTP request/response.
```
Comment (whoisj) [inline components/inference-service/requirements.md:21]: this is my mistake, I gave you an incorrect URL.

here is the correct URL: https://github.com/Project-MONAI/monai-deploy/blob/main/guidelines/monai-workloads.md#synchronous-computational-workload

File: components/inference-service/requirements.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+### Deployable on MONAI Operating Environments
+MIS SHALL run on Staging Server and Production Server environments as defined in [MONAI Operating Environments](https://github.com/Project-MONAI/monai-deploy/blob/main/guidelines/MONAI-Operating-Environments.md).
+
+### API Interface
+MIS SHALL provide a REST based RPC API which utilizes the functionality of the HTTP transport.
+
+### Consistent and Robust Logging
+MIS SHALL provide consistent and robust logging about it's operations.
```
Comment (whoisj) [inline components/inference-service/requirements.md:30]: grammar: `it's` => `its`

File: components/inference-service/requirements.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+### Register single MAP configuration before MIS startup
+MIS SHALL allow Clients to register the configuration of one MAP before MIS is started.
```
Comment (whoisj) [inline components/inference-service/requirements.md:33]: nit: `Clients` => `clients`

and maybe replace `register the configuration of one MAP before MIS is started.` with `provide MAP configuration as part of MIS' deployment configuration.`

File: components/inference-service/requirements.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+### Fulfill an inference request with uploaded dataset
+MIS SHALL fulfill an inference request with uploaded input dataset.
+
+### Fulfill one inference request at a time
+MIS SHALL run only fulfill one inference request at a time.
```
Comment (whoisj) [inline components/inference-service/requirements.md:39]: again - limitation, not requirement.

if we could fulfill multiple in parallel we would (and we maybe able to in v0.2 or later)

File: components/inference-service/requirements.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+### Provision resources for an inference request
+MIS SHALL provision CPU, memory, and GPU resources for an inference request as defined in the configuration.
```
Comment (whoisj) [inline components/inference-service/requirements.md:42]: I think requirements requires another requirement about MIS utilizing Kubernetes resource management (limits, etc.)

File: components/inference-service/requirements.md [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+### Provide status of inference request
+MIS SHALL provide status of inference request as a part of the response to the request.
```
Comment (whoisj) [inline components/inference-service/requirements.md:45]: how? do you mean the HTTP Error Code?
that's required by using HTTP and is covered by `MIS SHALL provide a REST based RPC API which utilizes the functionality of the HTTP transport.`

do you mean something else/different?

**Review by whoisj — CHANGES_REQUESTED** [review]
Thanks for keeping up with this, and apologies for the all the change requests.

I do think we need to differentiate between limitations (long-term and near-term) and requirements.

File: components/inference-service/requirements.md (hunk 8 above)
Comment (CPBridge) [inline components/inference-service/requirements.md:39]: Not really sure why this is a requirement. Is this just v1?

File: components/inference-service/requirements.md [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+MIS SHALL provide a REST based RPC API which utilizes the functionality of the HTTP transport.
+
+### Consistent and Robust Logging
+MIS SHALL provide consistent and robust logging about it's operations.
+
+### Register single MAP configuration before MIS startup
+MIS SHALL allow Clients to register the configuration of one MAP before MIS is started.
+
+### Fulfill an inference request with uploaded dataset
+MIS SHALL fulfill an inference request with uploaded input dataset.
```
Comment (CPBridge) [inline components/inference-service/requirements.md:36]: Again, "dataset" is vague. Imaging study/series? Does uploaded mean that the input data are in the payload of the request? I'm not sure that that's the best way to ingest input data, probably it should be pulled from a PACS directly

File: components/inference-service/requirements.md (hunk 5 above)
Comment (CPBridge) [inline components/inference-service/requirements.md:21]: Some of the workloads outlined in that document cannot be completed within a single HTTP request (e.g. large scale batch inference). Is this the intention:
```suggestion
### Support for Specific MONAI Workloads
MIS SHALL support those [MONAI workloads](https://github.com/Project-MONAI/monai-deploy/blob/main/guidelines/monai-workloads.md#clinical-use-cases) that can be completed within the timeframe of a single HTTP request/response.
```

File: components/inference-service/requirements.md (hunk 6 above)
Comment (CPBridge) [inline components/inference-service/requirements.md:30]:
```suggestion
MIS SHALL provide consistent and robust logging about its operations.
```
Should there be some requirement about error handling of exceptions that occur during the execution of the MAP. What is returned in this scenario?

File: components/inference-service/requirements.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+MIS SHALL track job execution progress, completion, and failures.
+
+### Provide status of completed job
+MIS SHALL allow Clients to retrieve status of run job as a part of the response of job submission request.
+
+### Provide results of completed job
+MIS SHALL allow Clients to download results of completed jobs as a part of the response of job submission request.
+
+### Cleanup inputs associated with a completed job
+MIS SHALL cleanup inputs and outputs associated with a job that has finished running.
```
[the remark this answers is not in the document]
Comment (dbericat) [reply components/inference-service/requirements.md]: +1. We should add them to the Backlog though, for 0.2 or 0.3

File: components/inference-service/requirements.md (hunk 9 above)
Comment (pritishnahar95) [reply components/inference-service/requirements.md:42]: That would be implementation details, isn't it?

File: components/inference-service/requirements.md (hunk 8 above)
Comment (pritishnahar95) [reply components/inference-service/requirements.md:39]: Moving from requirement to a v1 limitation.

File: components/inference-service/requirements.md (hunk 11 above)
Comment (pritishnahar95) [reply components/inference-service/requirements.md:36]: We have Informatics Gateway to ingest data from PACs. There is then a Workflow Manager which uploads this data ingested data from PACs to Inferencing Service.

File: components/inference-service/requirements.md (hunk 5 above)
Comment (pritishnahar95) [reply components/inference-service/requirements.md:21]: Updated workloads

**Pull request merged**